README.md
# learn-clojure
The early days of learning clojure. :)


basics/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-07-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-07-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/basics/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/basics/compare/0.1.0...0.1.1


basics/test/basics/core_test.clj
(ns basics.core-test
  (:require #_[clojure.test :refer :all]
            [basics.core :refer :all]
            [basics.my_functs :refer :all]
            [expectations :as exlib]))


;; Sam concat
(exlib/expect [1 2 3] (sam-cat [1 2] [3]))
(exlib/expect [1 2 3] (sam-cat [] [1 2 3]))
(exlib/expect [7 8 9] (sam-cat [7 8 9] []))
(exlib/expect [] (sam-cat [] []))

;; Sam Map
(exlib/expect '(4 5 6) (sam-map inc [3 4 5]))
(exlib/expect '(1 2 3 4) (sam-map dec [2 3 4 5]))
(exlib/expect '(4 3 2 1) (sam-map #(/ % 4) [16 12 8 4]))

;; Sam Map with helper instead of concat
(exlib/expect '(4 3 2 1) (sam-mcc #(/ % 4) [16 12 8 4] []))
(exlib/expect '() (sam-mcc inc [] []))

;; Sam Reduce
(exlib/expect 10 (sam-red + '(1 2 3 4) 0))
(exlib/expect 11 (sam-red #(+  (* 2 %1) %2) [1 2 3] 0))

;; Sam Drop
(exlib/expect [2 3] (sam-drop 1 [1 2 3]))
(exlib/expect [1 2 3] (sam-drop -1 [1 2 3]))
(exlib/expect '() (sam-drop 5 [1 2 3]))

;; Sam take
(exlib/expect [1] (sam-take 1 [1 2 3]))
(exlib/expect [1 2 3] (sam-take 5 [1 2 3]))
(exlib/expect '() (sam-take -1 [1 2 3]))

;; MultiArity Func
(exlib/expect 4 (tp 3))
(exlib/expect 5 (tp 3 2))

;; Sam take-while
(exlib/expect '() (sam-take-while #(> % 5) [1 2 3 4 3 2 1]))
(exlib/expect '(6 7 8) (sam-take-while #(> % 5) [6 7 8 3 2 1]))
(exlib/expect '(1 2 3 4) (sam-take-while #(> 5 %) [1 2 3 4 5 6 7 8 9 10]))

;; Sam drop-while
(exlib/expect '() (sam-drop-while #(> 5 %) [1 2 3 4 3 2 1]))
(exlib/expect '(3 2 1) (sam-drop-while #(> % 5) [6 7 8 3 2 1]))
(exlib/expect '(5 6 7 8 9 10) (sam-drop-while #(> 5 %) [1 2 3 4 5 6 7 8 9 10]))

;; Sam filter
(exlib/expect '(2 3 4 3 2) (sam-filter #(not= 1 %) [1 2 3 4 3 2 1]))
(exlib/expect '(6 7 8) (sam-filter #(> % 5) [6 7 8 3 2 1]))
(exlib/expect '(2 4 6 8 10) (sam-filter even? [1 2 3 4 5 6 7 8 9 10]))

;; Sam some
(exlib/expect true (sam-some odd? [2 4 6 1]))
(exlib/expect nil (sam-some odd? [2 4 6 8]))
(exlib/expect nil (sam-some even? []))

;; Sam Remove first element
(exlib/expect '(1 3 4 3 2 1) (mod-remove #(= 2 %) [1 2 3 4 3 2 1]))
(exlib/expect '(6 7 8 3 2 1) (mod-remove #(= 9 %) [6 7 8 3 2 1]))
(exlib/expect '(3 3) (mod-remove #(= 3 %) [3 3 3]))

;; Sam Sort
(exlib/expect [1 2 2 3 4 5 5] (sam-sort [5 1 3 2 5 2 4]))
(exlib/expect [1 2 3 4] (sam-sort [4 3 2 1]))

;; Sam min helper
(exlib/expect 3 (sam-min-help [7 6 5 4 3] 7))
(exlib/expect 1 (sam-min-help [1 2 3 4] 1))

;; Sam min
(exlib/expect 1 (sam-min 4 3 2 1))
(exlib/expect 1 (sam-min 4 3 7 8 9 2 41 52 231 1))

;; Sam add-one
(exlib/expect [1 2 3] (add-one [1 2] [3]))
(exlib/expect [1 2 3 4 5 6] (add-one [] [1 2 3 4 5 6]))
(exlib/expect [1 2 3] (add-one [1 2 3] []))

;; Sam all concat helper
(exlib/expect [1 2 3 4 5 6] (all-help [] '([1 2] [3 4] [5 6])))
(exlib/expect [1 2 3 4 5 6] (all-help [] '([] [1 2 3 4] [5 6])))
(exlib/expect [1 2 3 4 5 6] (all-help [] '([1 2 3] [] [4 5 6])))

;; Sam concat n no of lists
(exlib/expect [1 2 3 4 5 6] (all-cat [1 2] [3 4] [5 6]))
(exlib/expect [1 2 3 4 5 6] (all-cat [] [1 2 3 4] [5 6]))
(exlib/expect [1 2 3 4 5 6] (all-cat [1 2 3] [] [4 5 6]))

;; Sam into
(exlib/expect [1 2 3] (sam-into [] '(1 2 3)))
(exlib/expect '(3 2 1) (sam-into '() [1 2 3]))
(exlib/expect #{1 2 3} (sam-into #{} [1 2 3]))

;; Sam partial
(exlib/expect 14 (add2 4 5))

;; Sam min helper
(exlib/expect ["a" 1] (cus-min-help {"a" 1, "ccc" 3, "d" 1, "bb" 2} ["a" 1]))

;; Sam min from hash map
(exlib/expect ["a" 1] (cus-min {"dd" 2, "bbb" 3, "cccc" 4, "a" 1}))

;; Sam pair creates hash map of key and value after applying funcs to each ele
(exlib/expect {"dd" 2, "bbb" 3, "a" 1, "cc" 2} (sam-pair count ["dd" "bbb" "a" "cc"] {}))

;; Sam unpair hash map
(exlib/expect ["a" "bb" "ccc" "dddd"] (sam-unpair '(["a" 1] ["bb" 2] ["ccc" 3] ["dddd" 4]) []))

;; Sam cus remove one from hash map
(exlib/expect '(["aa" 2] ["bbbb" 4] ["ddd" 3]) (cus-remove #(= "b" %) '(["aa" 2] ["bbbb" 4] ["ddd" 3] ["b" 1]) ))

;; Sam cus sort key value pair
(exlib/expect '(["a" 1] ["dd" 2] ["bbb" 3] ["cccc" 4] ) (cus-sort {"dd" 2, "bbb" 3, "cccc" 4, "a" 1}))

;; Sam sort by
(exlib/expect '("a" "c" "bb" "ddd") (sam-sortby count ["bb" "a" "ddd" "c"]))
(exlib/expect '("a" "bb" "cc" "ddd" "eee") (sam-sortby count ["bb" "a" "ddd" "eee" "cc"]))

(exlib/expect 1400 (exlib/in [100 200 300 1400]))

(exlib/expect  {1  {:pegged true, :connections {6 3, 4 2}},
                2  {:pegged true, :connections {9 5, 7 4}},
                3  {:pegged false, :connections {10 6, 8 5}},
                4  {:pegged false, :connections {13 8, 11 7, 6 5, 1 2}},
                5  {:pegged false, :connections {12 8, 14 9}},
                6  {:pegged false, :connections {15 10, 13 9, 4 5, 1 3}},
                7  {:pegged false, :connections {9 8, 2 4}},
                8  {:pegged false, :connections {10 9, 3 5}},
                9  {:pegged true, :connections {7 8, 2 5}},
                10 {:pegged false, :connections {8 9, 3 6}},
                11 {:pegged false, :connections {13 12, 4 7}},
                12 {:pegged false, :connections {14 13, 5 8}},
                13 {:pegged false, :connections {15 14, 11 12, 6 9, 4 8}},
                14 {:pegged false, :connections {12 13, 5 9}},
                15 {:pegged false, :connections {13 14, 6 10}},
                :rows 5} (mymap) )

(exlib/expect {1 {:pegged true, :connections {6 3, 4 2}}, 2 {:pegged true, :connections {9 5, 7 4}}, 9 {:pegged true, :connections {7 8, 2 5}}} (gettrue (mymap)))

(exlib/expect {6 3, 4 2} (getconnects (gettrue (mymap)) 1))


basics/test/basics/proj_eul_test.clj
(ns basics.proj_eul_test
  (:require [basics.proj_euler :refer :all]
            [clojure.math.numeric-tower :as math]
            [expectations :as exlib]))

;; Prob 1
(exlib/expect 23 (sumofmul 10))
(exlib/expect 45 (sumofmul 15))

;; Prob 2
(exlib/expect 1 (fib 2))
(exlib/expect 2 (fib 3))
(exlib/expect 3524578 (fib 33))
(exlib/expect 4613732 (sumfib 33))

;; Prob 3
(exlib/expect '(2 3 5 6 10 15) (fact 30 2 []))
(exlib/expect true  (prime? 29 2))
(exlib/expect false  (prime? 30 2))
#_(exlib/expect 6857 (largeprime 600851475143))

;; Prob 4
(exlib/expect true (palindrome? 909))

;; Prob 5
(exlib/expect false (even_div? 2520))
(exlib/expect false (even_div? 5000))
(exlib/expect true (even_div? 2432902008176640000))

;; Prob 6
#_(exlib/expect 2640 (diff_sum_sqr)) ;; for 10
#_(exlib/expect 41230 (diff_sum_sqr)) ;; for 20
#_(exlib/expect 25164150 (diff_sum_sqr)) ;; for 100

;; Prob 7
(exlib/expect true (prime? 2 2))
#_(exlib/expect 104743 (nthprime 0 0 2))
#_(exlib/expect 29 (nthprime 0 0 2)) ;; for 10

;; Prob 10
#_(exlib/expect 17 (sumprime 0 2)) ;; 10
;;(exlib/expect 142913828922 (sumprime 0 2)) ;; 10001

;; Prob 8
(exlib/expect '(1 2 3 4 5 6) (digits 123456))
(exlib/expect 24 (mulall [1 2 3 4]))
#_(exlib/expect 5832 (highmul (digits in) 0)) ;;  for 4digit product

;; Prob 12


basics/src/basics/core.clj
(ns basics.core
  (:gen-class))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

(defn take-input []
  "Accept user input"
  (println "Enter a input")
  (let [inp (read-line)]
    (println inp)))

(defn make-board []
  "Static map of length 5"
  {1  {:pegged false, :connections {6 3, 4 2}},
   2  {:pegged true, :connections {9 5, 7 4}},
   3  {:pegged true, :connections {10 6, 8 5}},
   4  {:pegged true, :connections {13 8, 11 7, 6 5, 1 2}},
   5  {:pegged true, :connections {14 9, 12 8}},
   6  {:pegged true, :connections {15 10, 13 9, 4 5, 1 3}},
   7  {:pegged true, :connections {9 8, 2 4}},
   8  {:pegged true, :connections {10 9, 3 5}},
   9  {:pegged true, :connections {7 8, 2 5}},
   10 {:pegged true, :connections {8 9, 3 6}},
   11 {:pegged true, :connections {13 12, 4 7}},
   12 {:pegged true, :connections {14 13, 5 8}},
   13 {:pegged true, :connections {15 14, 11 12, 6 9, 4 8}},
   14 {:pegged true, :connections {12 13, 5 9}},
   15 {:pegged true, :connections {13 14, 6 10}}})

(defn get-peg-value [inp board out]
  "Returns 1 if pegged else returns - for the given input"
  (if (empty? inp)
    out
    (recur (rest inp) board (conj out (if (= (get-in board [(first inp) :pegged]) true)
                                      1
                                      0)))))



(defn get-valid-pegs [inp]
  "Gets the filled pegs"

  (keys  (filter (fn [[k v]] (:pegged v)) inp)))

(defn convert-to-alph
  "Coverts filled pegs numbers to their respective alphabets"
  ([inp]
   (convert-to-alph inp [] ))
  ([inp out]
   (if (empty? inp)
     out
     (recur (rest inp) (conj out (char (+ 96 (first inp))))))))

(defn convert-to-num
  "Converts user-accepted alphabet to respective number"
  ([inp]
   (convert-to-num inp []))
  ([inp out]
   (if (empty? inp)
     (filter pos? out)
     (recur (rest inp) (conj out (- (int (first inp)) 96))))))
(defn tri*
  "Generates lazy sequence of triangular numbers"
  ([] (tri* 0 1))
  ([sum n]
   (let [new-sum (+ sum n)]
     (cons new-sum (lazy-seq (tri* new-sum (inc n)))))))

(def tri (tri*))

(defn get-element [inp]
  "Gets the elements on the respective level"

  (let [drop-value (nth (conj (take 5 tri) 0) (dec inp))]
    (take inp (drop drop-value (range 1 16)))))

#_(defn convert-zip-format-1
  "Converts to vector of strings EG: '(4 5 6) -> [\"4\" \"5\" \"6\"] "
  ([inp]
   (convert-zip-format inp []))
  ([inp out]
   (if (empty? inp)
     out
     (recur (rest inp) (conj out (str (first inp)))))))

(defn convert-zip-format [inp]
  (vec inp))

(defn comb-element-peg [inp board]
  ""
  (let [a (get-element inp)
        b (get-peg-value a board [])]
    (zipmap (convert-zip-format (convert-to-alph  a)) (convert-zip-format b))))

(defn print-map [board]
  "Prints the board"
  (loop [cl 1
         ce 1
         ct 4]
    (when (<= cl 5)
      (println (str (apply str (repeat ct "\t")) (comb-element-peg ce board)))
      (recur (inc cl) (inc ce) (dec ct) ))))


;; RADHA CODE

(defn char-as-str [x] (filter (comp not empty?) (clojure.string/split x #" ")))

(defn convert-int [str] (- (int (let [gb (.getBytes str)] (char (first gb)))) 96))

                         ;taking move
;(char-as-str "a   b");is peg

(defn peg? [board pos] (if (get-in board [pos :pegged])
                        true
                        false));valid move;remove peg

(defn remove-peg [board pos]
 (assoc-in board
           [pos :pegged]
           false))
                                        ;add peg

(defn add-peg [board pos]
 (assoc-in board
           [pos :pegged]
           true))
                                        ;move peg

(defn neigh-peg [board pos dest]
  (second (first (filter #(= dest (first %))
                         (get-in board
                                 [pos :connections])))))


(defn move-peg [board pos dest]
 (let [neigh (neigh-peg board pos dest)]
   ;(println neigh)
   (add-peg (remove-peg (remove-peg board
                                neigh)
                    pos)
            dest)))

                                        ;is valid

(defn valid-move? [board pos dest]
 (let [neigh (neigh-peg board pos dest)]
   (and (peg? board pos)
        (peg? board neigh)
        (not (peg? board dest)))))

(defn take-input [board]
  (println "enter the source and destination points")
  (let [lis (map convert-int  (char-as-str (read-line)))
        fir (first lis)
        sec (second lis)]
    (if (valid-move? board fir sec)
      (move-peg board fir sec)
      (do println "invalid move "
          board))))


;;HItesh COde
(comment
  (defn checkres [board]
    (let [truemap (filter #(true? (:pegged (second %))) (vec board))]
      (if (= 1 (count truemap))
        (println "WINNER")

        (if (movava truemap)
          (println "U LOST")
          #_(println "U LOST"))) ))

  (defn ispeg [board pos]
    (:pegged (second (first (filter #(= pos (first %)) board)))))

  (defn movava [board]
    #_(let [newboard (filter #(true? (:pegged (second %))) (vec board))])
    (and
     (false? (ispeg board (first (first (:connections (second (first board)))))))
     (true? (ispeg board (second (first (:connections (second (first board))))))))))



(defn mymap []  {1  {:pegged true, :connections {6 3, 4 2}},
                2  {:pegged true, :connections {9 5, 7 4}},
                3  {:pegged false, :connections {10 6, 8 5}},
                4  {:pegged false, :connections {13 8, 11 7, 6 5, 1 2}},
                5  {:pegged false, :connections {12 8, 14 9}},
                6  {:pegged false, :connections {15 10, 13 9, 4 5, 1 3}},
                7  {:pegged false, :connections {9 8, 2 4}},
                8  {:pegged false, :connections {10 9, 3 5}},
                9  {:pegged true, :connections {7 8, 2 5}},
                10 {:pegged false, :connections {8 9, 3 6}},
                11 {:pegged false, :connections {13 12, 4 7}},
                12 {:pegged false, :connections {14 13, 5 8}},
                13 {:pegged false, :connections {15 14, 11 12, 6 9, 4 8}},
                14 {:pegged false, :connections {12 13, 5 9}},
                15 {:pegged false, :connections {13 14, 6 10}},
                :rows 5}
  )


(defn ispeg [board pos]
  (:pegged (second (first (filter #(= pos (first %)) board)))))

(defn gettrue [board]
  (into {} (filter #(true? (:pegged (second %))) board)))

(defn getconnects [board no]
  (get-in board [no :connections]))

#_(some (fn [x]
        ((println x)
         (and
          (false? (ispeg board (first x)))
          (true? (ispeg board (second x))))))
      ((println (first %))
       (getconnects board (first %))))

(defn check [board]
  (map #(getconnects board (first %))
       board))

(defn checkallconnections [board conn]
  (some (fn [x]
          (and
           (false? (peg? board (first x)))
           (true? (peg? board (second x)))))
        (vec conn)))

(defn movava [board]
  (let [newboard (gettrue board)]
    (println newboard)
    (some #(and
            (false? (ispeg board (first (first (:connections (second %))))))
            (true? (ispeg board (second (first (:connections (second %))))))
            ) newboard)))


(defn checkres [board]
  (let [truemap (gettrue board)]
    (if (= 1 (count truemap))
      (println "WINNER")
      (if (movava truemap)
        (println "U LOST")
        #_(println "U LOST"))) ))


basics/src/basics/proj_euler.clj
(ns basics.proj_euler
  (:gen-class)
  (:require [clojure.math.numeric-tower :as math]))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

;; Prob 1 Sum of multiples of 3 & 5
(defn sumofmul [n]
  (reduce + (filter #(or (zero? (mod % 3)) (zero? (mod % 5))) (range 1 n))))

;; Prob 2 Even Fibonacci No
(defn fib [n]
  (if (zero? n)
    0
    (if (= 1 n)
      1
      (+ (fib (- n 1)) (fib (- n 2))))))

(defn sumfib [n]
  (reduce + (filter #(and (<= % 4000000) (even? %)) (map #(fib %) (range 2 (+ n 2))))))

;; Prob 3
(defn fact [n c faclis]
  (if (> c (/ n 2))
    faclis
    (if (zero?  (mod n c))
      (fact n (inc c)  (conj faclis c))
      (fact n (inc c) faclis)
      )))

(defn prime? [x c]
  (if (> c (math/sqrt x))
    true
    (if (zero? (mod x c))
      false
      (recur x (inc c)))))

(defn largeprime [n]
  (last (filter #(prime? % 2) (fact n 2 []))))

;; Prob 4 Largest palindrome product

(defn largepalin [x]
  x)

(defn palindrome? [n]
  (if (= (str n) (clojure.string/reverse (str n)))
    true
    false))

#_(defn lp
  (loop [x 999 y 999]
    (when (> y 100)
      ())
    )
    )

;; Prob 5
(defn even_div? [no]
  (= 20
     (count (take-while #(= 0(mod no %))
                        (range 1 21)))))

(defn smallest [x]
  (if (even_div? x)
    x
    (recur (inc x))))

;; Prob 6
(defn diff_sum_sqr []
  (let [arr (range 1 101)
        sumsqr (reduce + (map (fn [x] (* x x)) arr))
        sqrsum (* (reduce + arr) (reduce + arr))]
    (- sqrsum sumsqr)))

;; Prob 7
(defn nthprime [primecount primeno count]
  (if (= 10001 primecount)
    primeno
    (if (prime? count 2)
      (recur (inc primecount) count (inc count))
      (recur primecount primeno (inc count)))))

;; Prob 8
(def in 7316717653133062491922511967442657474235534919493496983520312774506326239578318016984801869478851843858615607891129494954595017379583319528532088055111254069874715852386305071569329096329522744304355766896648950445244523161731856403098711121722383113622298934233803081353362766142828064444866452387493035890729629049156044077239071381051585930796086670172427121883998797908792274921901699720888093776657273330010533678812202354218097512545405947522435258490771167055601360483958644670632441572215539753697817977846174064955149290862569321978468622482839722413756570560574902614079729686524145351004748216637048440319989000889524345065854122758866688116427171479924442928230863465674813919123162824586178664583591245665294765456828489128831426076900422421902267105562632111110937054421750694165896040807198403850962455444362981230987879927244284909188845801561660979191338754992005240636899125607176060588611646710940507754100225698315520005593572972571636269561882670428252483600823257530420752963450)

(defn digits [number]
  (map #(Character/digit % 10) (str number)))

(defn mulall [arr]
  (apply * arr))

(defn highmul [ls prod]
  (if (< (count ls) 13)
    prod
    (let [curr  (mulall  (take 13 ls))]
      (if (> curr prod)
        (recur (rest ls) curr)
        (recur (rest ls) prod)
        ))))

;; Prob 9

;; Prob 10
(defn sumprime [primesum count]
  (if (= 2000000 count)
    primesum
    (if (prime? count 2)
      (recur (+ primesum count) (inc count))
      (recur primesum (inc count)))))


;; Prob 12


basics/src/basics/my_functs.clj
(ns basics.my_functs)

;; Custom implementation of inbuilt clojure functions

;; 1. Custom concat for only 2 args
(defn sam-cat [ls1 ls2]
  (if (and (empty? ls1) (empty? ls2))
    '()
    (if (empty? ls2)
      (seq ls1)
      (sam-cat (conj (vec ls1) (first ls2)) (rest ls2)))))

;; 2. Custom map
(defn sam-map [f arg]
  (when-not (empty? arg)
    (sam-cat
     [(f (first arg))]
     (sam-map f (rest arg)))))

;; 3. Custom map with helper instead of concat
(defn sam-mcc [f arr acc]
  (if (empty? arr)
    acc
    #_(sam-mcc f (rest arr) (conj acc (f (first arr)))) ;; Sir ka method add to acc
    #_(conj  (sam-mcc f (rest arr) acc) (f (first  arr))) ;;Gives Reverse : conj adds @ last
    (cons (f (first arr)) (sam-mcc f (rest arr) acc) ) ;; cons use karke hogaya :P
    ))

;; 4. Custom reduce
(defn sam-red [f arr acc]
  (if (empty? arr)
    acc
    (sam-red f (rest arr) (f acc (first arr)))))

;; 5. Custom drop
(defn sam-drop [n arr]
  (if (<= n 0)
    arr
    (sam-drop (dec n) (rest arr))))

;; 6. Custom take
(defn sam-take [n arr]
  (if (>= n (count arr))
    arr
    (if (<= n 0)
      '()
      (conj (sam-take (dec n) (rest arr)) (first arr)))))

;; 7. MultiArity function
(defn tp
  ([a] (inc a))
  ([b c] (+ b c)))

;; 8. Custom take-while
(defn sam-take-while [f arr]
  (if (empty? arr)
    arr
    (if (f (first arr))
      (conj (sam-take-while f (rest arr)) (first arr))
      '())))

;; 9. Custom drop-while
(defn sam-drop-while [f arr]
  (if (empty? arr)
    arr
    (if (f (first arr))
      (sam-drop-while f (rest arr))
      arr )))

;; 10. Custom filter
(defn sam-filter [f arr]
  (when-not (empty? arr)
    (if (f (first arr))
      (cons (first arr) (sam-filter f (rest arr)))
      (sam-filter f (rest arr)))))

;; 11. Custom some
(defn sam-some [f arr]
  (when-not (empty? arr)
    (if (f (first arr))
      true
      (sam-some f (rest arr)))))

;; 12. Custom Remove first element
(defn mod-remove [f arr]
  (letfn ([helper [f in out]
           (if (or (empty? in)
                   (f (first in)))
             (concat out (rest in))
             (recur f (rest in) (conj out (first in))))])
    (helper f arr [])))

;; 13. Custom Sort
(defn sam-sort [arr]
  (when-not (empty? arr)
    (let [min-arr (apply min arr)]
      (conj (sam-sort (mod-remove #(= min-arr %)
                                  arr))
            min-arr))))

;; 14. Custom min/max For max replace < with >
(defn sam-min-help [arr acc]
  (if (empty? arr)
    acc
    (recur (rest arr)
           (if (< acc (first arr))
             acc
             (first arr)))))


(defn sam-min [& args]
  (if (empty? (rest args))
    (first args)
    (sam-min-help (rest args) (first args))))

;; 15. Custom concat with any args
(defn add-one [acc ls]
  (if (empty? ls)
    acc
    (add-one (conj acc (first ls)) (rest ls))))

(defn all-help [acc arr]
  (if (empty? arr)
    acc
    (all-help (add-one acc (first arr))
              (rest arr))))

(defn all-cat [& args]
 #_ (all-help [] args) ;; with use of helper
  (reduce add-one args) ;; when we want to reduce to one use it
  #_(if (empty? (first (rest args))) ;; gaya hua logic
    (first args)
    (all-cat (add-one (first args) (second args)) (drop 2 args))))

;; 16. Custom into
(defn sam-into [out in]
  (apply conj out in))

;; 17. Custom complement
(defn sam-comp [f]
  (fn [& args]
    (not (apply f args))))

;; 18. Custom partial
(defn sam-par [f & arg]
  (fn [& args]
    (apply f (into arg args))))

(defn add [a b c d]
  (+ a b c d))

(def add2 (sam-par add 2 3))

;; 19. Custom sort-by
(defn cus-min-help [arr acc]
  (if (empty? arr)
    acc
    (recur (rest arr)
           (if (<= (second acc) (second (first arr)))
             acc
             (first arr)))))

(defn cus-min [args]
  (if (empty? (rest args))
    (first args)
    (cus-min-help (rest args) (first args))))

(defn sam-pair [f arr acc]
  (if (empty? arr)
    acc
    (sam-pair f
          (rest arr)
          (assoc acc (first arr) (f (first arr))))))

(defn sam-unpair [arr acc]
  (if (empty? arr)
    acc
    (sam-unpair (rest arr)
                (conj  acc (first (first arr))))))

(defn cus-remove [f arr]
  (letfn ([helper [f in out]
           (if (or (empty? in)
                   (f (first (first in))))
             (concat out (rest in))
             (recur f (rest in) (conj out (first in))))])
    (helper f arr [])))

(defn cus-sort [arr]
  (when-not (empty? arr)
    (let [min-arr (cus-min arr)]
      (conj (cus-sort (cus-remove #(= (first min-arr) %)
                                  arr))
            min-arr))))

(defn sam-sortby [f arr]
  (let [key-val-arr (sam-pair f arr {})
        sorted-pair (cus-sort key-val-arr)]
    (sam-unpair sorted-pair [])))


;; 20. Replacing a string in an passed string and returning the modified string
(require '[clojure.string :as s])

(defn clean
  [text]
  (s/replace (s/trim text) #"lol" "LOL"))

;;(clean "Are u nuts..!! lol! ;)  ")

;; 21. Sam reverse
(defn sam-rev [ls rev]
  (if (empty? ls)
    rev
    (sam-rev (rest ls) (cons (first ls) rev ))))

;; 22. Sam Complement

;; Sam comp helper
(defn comp-help [fls acc]
  (if (empty? fls)
    acc
    (comp-help (rest fls) ((first fls) acc))))

;; Sam comp
(defn sam-comp [& funcs]
  (let [revf (sam-rev funcs [])]
    (fn f [& args]
      (comp-help (rest revf) (apply (first revf) args)))))
